swift/ql/test/extractor-tests/generated/decl/CapturedDecl/closures.swift

```diff
@@ -1,26 +1,211 @@
-func bar() -> String {
+func hello() -> String {
   return "Hello world!"
 }
 
-func foo() {
+func captureList() {
   let y = 123
-  { [x = bar()] () in
+  { [x = hello()] () in
      print(x)
      print(y) }()
 }
 
 var escape: (() -> ())? = nil
 
-func baz() {
+func setEscape() {
   var x = 0
-  func quux() {
+  escape = {
     x += 1
     print(x)
   }
-  escape = quux
 }
 
 func callEscape() {
-  baz()
+  setEscape()
   escape?()
 }
+
+func logical() -> Bool {
+  let f: ((Int) -> Int)? = { x in x + 1 }
+  let x: Int? = 42
+  return f != nil
+      && (x != nil
+          && f!(x!) == 43)
+}
+
+func asyncTest() {
+  func withCallback(_ callback: @escaping (Int) async -> Int) {
+    @Sendable
+    func wrapper(_ x: Int) async -> Int {
+      return await callback(x + 1)
+    }
+    Task {
+      print("asyncTest():", await wrapper(40))
+    }
+  }
+  withCallback { x in
+    x + 1
+  }
+}
+
+func foo() -> Int {
+  var x = 1 // x is a non-escaping capture of f and r
+  let f = { y in x += y }
+  x += 40
+  let r = { x }
+  f(1)
+  return r() // 42
+}
+
+func bar() -> () -> Int {
+  var x = 1 // x is a non-escaping capture of f, escaping capture of r
+  let f = { y in x += y }
+  x += 40
+  let r = { x }
+  f(1)
+  return r // constantly 42
+}
+
+var g: ((Int) -> Void)? = nil
+func baz() -> () -> Int {
+  var x = 1 // x is an escaping capture of g and r
+  g = { y in x += y } // closure escapes!
+  x += 40
+  let r = { x }
+  g!(1)
+  return r
+}
+
+func quux() -> Int {
+  var y = 0
+
+  func f() -> () -> Void {
+    var x = 5
+
+    func a() {
+      y = 10*y + x
+      x -= 1
+      if x > 0 {
+        b()
+      }
+    }
+
+    func b() {
+      y = 10*y + 2*x
```

Comment:
Why are there two results for `y` on line 85:
```
| closures.swift:85:7:85:7 | y | getModule: | file://:0:0:0:0 | closures | getNumberOfMembers: | 0 | getDecl: | closures.swift:79:7:79:7 | y | isDirect: | no | isEscaping: | yes |
| closures.swift:85:7:85:7 | y | getModule: | file://:0:0:0:0 | closures | getNumberOfMembers: | 0 | getDecl: | closures.swift:79:7:79:7 | y | isDirect: | yes | isEscaping: | yes |
```
but only one result for `y` on (identical) line 93?
```
| closures.swift:93:7:93:7 | y | getModule: | file://:0:0:0:0 | closures | getNumberOfMembers: | 0 | getDecl: | closures.swift:79:7:79:7 | y | isDirect: | yes | isEscaping: | yes |
```
(I haven't entirely got to grips with what `isDirect` and `isEscaping` mean)

Reply:
AIUI, escaping = used by more than one closure, and indirect = captured but only for use in a sub-closure.

So the indirect capture should be `f`'s capture of `y`.

Reply:
Oh I see - I was fixated with `a` and `b` and didn't notice that line 85 is the first use of `y` in `f` as well as in `a`.

Thanks for the explanations.

:+1: